Support Scala 3 syntax

Running Stryker4s on a project that is using Scala 3 syntax will give a parsing error. For example, using the new enums:

```
[error] Error while parsing file 'src/main/scala/chapter4/Option.scala', expected class or object definition identifier
[error] /project/src/main/scala/Option.scala:3: error: expected class or object definition identifier
[error] enum Option[+A] {
[error] ^
[error] 	at flatMap @ stryker4s.Stryker4s.run(Stryker4s.scala:14)
[error] 	at unsafeRunSync @ stryker4s.sbt.Stryker4sMain$.$anonfun$strykerTask$1(Stryker4sMain.scala:47)
[error] 	at unsafeRunSync @ stryker4s.sbt.Stryker4sMain$.$anonfun$strykerTask$1(Stryker4sMain.scala:47)
[error] 	at traverse @ stryker4s.run.Stryker4sRunner.resolveReporters(Stryker4sRunner.scala:40)
[error] 	at traverse @ stryker4s.run.Stryker4sRunner.resolveReporters(Stryker4sRunner.scala:40)
[error] 	at traverse @ stryker4s.run.Stryker4sRunner.resolveReporters(Stryker4sRunner.scala:40)
[error] 	at flatMap @ stryker4s.run.Stryker4sRunner.run(Stryker4sRunner.scala:25)
[error] 	at map @ stryker4s.sbt.Stryker4sMain$.$anonfun$strykerTask$1(Stryker4sMain.scala:47)
[error] (stryker) /project/src/main/scala/Option.scala:3: error: expected class or object definition identifier
```

This is because the default parsing 'dialect' of Scalameta is 2.13. We should pass the proper [scala dialect](https://github.com/scalameta/scalameta/blob/master/scalameta/dialects/shared/src/main/scala/scala/meta/dialects/package.scala) to Scalameta. The most user-friendly way to do this would be to detect this automatically. This might only be possible in the sbt plugin. There should also be a config option (`scala-dialect`?) which would take precedence over the auto-selected setting. If nothing is set the default can still be 2.13 (which is also the Scalameta default and should work with all pre-scala 3 versions)

There might also be new mutations possible in Scala 3. This issue can also be used to track those.

--- a/.scalafmt.conf
+++ b/.scalafmt.conf
@@ -7,3 +7,6 @@ rewrite.rules = [SortImports, SortModifiers, RedundantParens]
 assumeStandardLibraryStripMargin = true
 
 onTestFailure = "To fix this, run './bin/scalafmt' from the project root directory"
+
+# Syntax is backwards-compatible and we also need to be able to parse scala 3 files
+runner.dialect = dotty

--- a/core/src/main/scala/stryker4s/config/Config.scala
+++ b/core/src/main/scala/stryker4s/config/Config.scala
@@ -3,6 +3,7 @@ package stryker4s.config
 import java.util.concurrent.TimeUnit
 
 import scala.concurrent.duration.FiniteDuration
+import scala.meta.{dialects, Dialect}
 
 import better.files._
 
@@ -18,7 +19,8 @@ final case class Config(
     timeout: FiniteDuration = FiniteDuration(5000, TimeUnit.MILLISECONDS),
     timeoutFactor: Double = 1.5,
     maxTestRunnerReuse: Option[Int] = None,
-    legacyTestRunner: Boolean = false
+    legacyTestRunner: Boolean = false,
+    scalaDialect: Dialect = dialects.Scala3
 )
 
 object Config {

--- a/core/src/main/scala/stryker4s/config/implicits/ConfigReaderImplicits.scala
+++ b/core/src/main/scala/stryker4s/config/implicits/ConfigReaderImplicits.scala
@@ -2,6 +2,9 @@ package stryker4s.config.implicits
 
 import java.nio.file.Path
 
+import scala.meta.Dialect
+import scala.meta.dialects._
+
 import better.files.File
 import pureconfig.ConfigReader
 import pureconfig.error.CannotConvert
@@ -40,34 +43,57 @@ trait ConfigReaderImplicits {
 
   implicit private[config] val uriReader = pureconfig.module.sttp.reader
 
-  implicit private[config] val thresholdsReader: ConfigReader[Thresholds] = deriveReader[Thresholds] emap {
-    case Thresholds(high, _, _) if isNotPercentage(high)   => notPercentageError(high, "high")
-    case Thresholds(_, low, _) if isNotPercentage(low)     => notPercentageError(low, "low")
-    case Thresholds(_, _, break) if isNotPercentage(break) => notPercentageError(break, "break")
-    case Thresholds(high, low, _) if high < low =>
-      Left(
-        CannotConvert(
-          high.toString(),
-          "thresholds.high",
-          s"'high' ($high) must be greater than or equal to 'low' ($low)"
+  implicit private[config] val thresholdsReader: ConfigReader[Thresholds] = {
+    def isNotPercentage(n: Int) = n < 0 || n > 100
+
+    def notPercentageError(value: Int, name: String): Left[CannotConvert, Thresholds] =
+      Left(CannotConvert(value.toString(), s"thresholds.$name", "must be a percentage 0-100"))
+
+    deriveReader[Thresholds] emap {
+      case Thresholds(high, _, _) if isNotPercentage(high)   => notPercentageError(high, "high")
+      case Thresholds(_, low, _) if isNotPercentage(low)     => notPercentageError(low, "low")
+      case Thresholds(_, _, break) if isNotPercentage(break) => notPercentageError(break, "break")
+      case Thresholds(high, low, _) if high < low =>
+        Left(
+          CannotConvert(
+            high.toString(),
+            "thresholds.high",
+            s"'high' ($high) must be greater than or equal to 'low' ($low)"
+          )
         )
-      )
-    case Thresholds(_, low, break) if low <= break =>
-      Left(
-        CannotConvert(
-          low.toString(),
-          "thresholds.low",
-          s"'low' ($low) must be greater than 'break' ($break)"
+      case Thresholds(_, low, break) if low <= break =>
+        Left(
+          CannotConvert(
+            low.toString(),
+            "thresholds.low",
+            s"'low' ($low) must be greater than 'break' ($break)"
+          )
         )
-      )
-    case valid => Right(valid)
+      case valid => Right(valid)
+    }
   }
 
-  private def isNotPercentage(n: Int) = n < 0 || n > 100
-
-  /** Helper function to create a CannotConvert when a value is not a percentage
-    */
-  private def notPercentageError(value: Int, name: String): Left[CannotConvert, Thresholds] =
-    Left(CannotConvert(value.toString(), s"thresholds.$name", "must be a percentage 0-100"))
+  implicit private[config] val dialectReader: ConfigReader[Dialect] = {
+    val scalaVersions = Map(
+      List("scala211", "scala2.11", "2.11", "211") -> Scala211,
+      List("scala212", "scala2.12", "2.12", "212") -> Scala212,
+      List("scala213", "scala2.13", "2.13", "213", "2") -> Scala213,
+      List("scala3", "scala3.0", "3.0", "3", "dotty") -> Scala3
+    )
+    ConfigReader[String].emap { input =>
+      scalaVersions
+        .collectFirst { case (strings, dialect) if strings.contains(input.toLowerCase()) => dialect }
+        .toRight(
+          CannotConvert(
+            input,
+            "scalaDialect",
+            s"Unsupported scalaDialect. Leaving this configuration empty defaults to scala3 which might also work for you. Valid scalaDialects are: ${scalaVersions
+              .flatMap(_._1)
+              .map(d => s"'$d'")
+              .mkString(", ")}."
+          )
+        )
+    }
+  }
 
 }

--- a/core/src/main/scala/stryker4s/mutants/findmutants/MutantFinder.scala
+++ b/core/src/main/scala/stryker4s/mutants/findmutants/MutantFinder.scala
@@ -1,8 +1,7 @@
 package stryker4s.mutants.findmutants
 
-import scala.meta.Parsed.{Error, Success}
-import scala.meta.Source
 import scala.meta.parsers.XtensionParseInputLike
+import scala.meta.{Dialect, Source}
 
 import better.files.File
 import stryker4s.config.Config
@@ -22,13 +21,18 @@ class MutantFinder(matcher: MutantMatcher)(implicit config: Config, log: Logger)
     (included.flatten, excluded.size)
   }
 
-  def parseFile(file: File): Source =
-    // Match is reported as non-exhaustive for some reason
-    (file.toJava.parse[Source]: @unchecked) match {
-      case Success(source) =>
-        source
-      case Error(_, msg, ex) =>
-        log.error(s"Error while parsing file '${file.relativePath}', $msg")
-        throw ex
-    }
+  def parseFile(file: File): Source = {
+    implicit val dialect: Dialect = config.scalaDialect
+
+    file.toJava
+      .parse[Source]
+      .fold(
+        e => {
+          log.error(s"Error while parsing file '${file.relativePath}', ${e.message}")
+          throw e.details
+        },
+        identity
+      )
+
+  }
 }

--- a/core/src/test/resources/scalaFiles/scala3File.scala
+++ b/core/src/test/resources/scalaFiles/scala3File.scala
@@ -1,0 +1,9 @@
+// Tests support for new top-level and extension syntax
+
+object SeqOps {
+  extension[A, B](list: Seq[A]) {
+    def myFoldRight(init: B)(f: (A, B) => B): B = list.foldRight(init)(f)
+  }
+}
+
+type Foo = String

--- a/core/src/test/scala/stryker4s/config/ConfigReaderTest.scala
+++ b/core/src/test/scala/stryker4s/config/ConfigReaderTest.scala
@@ -1,6 +1,7 @@
 package stryker4s.config
 
 import scala.concurrent.duration._
+import scala.meta.dialects._
 
 import better.files.File
 import pureconfig.ConfigSource
@@ -28,6 +29,7 @@ class ConfigReaderTest extends Stryker4sSuite with LogMatchers with ConfigReader
           config.timeout shouldBe 5.5.seconds
           config.maxTestRunnerReuse.value shouldBe 15
           config.legacyTestRunner shouldBe true
+          config.scalaDialect shouldBe Scala212
       }
     }
 
@@ -57,6 +59,7 @@ class ConfigReaderTest extends Stryker4sSuite with LogMatchers with ConfigReader
         version = None,
         module = None
       )
+      result.scalaDialect shouldBe Scala3
     }
 
     it("should fail on an empty config file") {
@@ -197,8 +200,52 @@ class ConfigReaderTest extends Stryker4sSuite with LogMatchers with ConfigReader
             case Right(value) => value shouldBe expected
             case Left(ConfigReaderFailures(ConvertFailure(reason, _, _), _*)) =>
               reason shouldBe expected
-            case other => fail(s"unexpected value $other")
+            case Left(other) => fail(s"unexpected value $other")
           }
+        }
+      }
+    }
+
+    describe("ScalaDialect") {
+      val validVersions = List(
+        "scala211" -> Scala211,
+        "scala2.11" -> Scala211,
+        "2.11" -> Scala211,
+        "211" -> Scala211,
+        "scala212" -> Scala212,
+        "scala2.12" -> Scala212,
+        "2.12" -> Scala212,
+        "212" -> Scala212,
+        "scala213" -> Scala213,
+        "scala2.13" -> Scala213,
+        "2.13" -> Scala213,
+        "213" -> Scala213,
+        "scala3" -> Scala3,
+        "dotty" -> Scala3,
+        "3" -> Scala3,
+        "3.0" -> Scala3,
+        "scala3.0" -> Scala3
+      )
+
+      validVersions.foreach { case (input, expected) =>
+        it(s"should parse $input to $expected") {
+          ExampleConfigs.scalaDialect(input).at("stryker4s").load[Config] match {
+            case Right(value) => value.scalaDialect shouldBe expected
+            case Left(value)  => fail(s"Expected valid parsing, got $value")
+          }
+
+        }
+      }
+
+      it("should not parse invalid scala-dialects") {
+        ExampleConfigs.scalaDialect("foobar").at("stryker4s").load[Config] match {
+          case Left(ConfigReaderFailures(ConvertFailure(reason, _, _), _*)) =>
+            reason shouldBe CannotConvert(
+              "foobar",
+              "scalaDialect",
+              "Unsupported scalaDialect. Leaving this configuration empty defaults to scala3 which might also work for you. Valid scalaDialects are: 'scala211', 'scala2.11', '2.11', '211', 'scala212', 'scala2.12', '2.12', '212', 'scala213', 'scala2.13', '2.13', '213', '2', 'scala3', 'scala3.0', '3.0', '3', 'dotty'."
+            )
+          case value => fail(s"Expected parsing failure but got $value")
         }
       }
     }

--- a/core/src/test/scala/stryker4s/mutants/findmutants/MutantFinderTest.scala
+++ b/core/src/test/scala/stryker4s/mutants/findmutants/MutantFinderTest.scala
@@ -9,6 +9,7 @@ import better.files.File
 import stryker4s.config.Config
 import stryker4s.extension.FileExtensions._
 import stryker4s.extension.TreeExtensions.IsEqualExtension
+import stryker4s.log.Logger
 import stryker4s.scalatest.{FileUtil, LogMatchers}
 import stryker4s.testutil.Stryker4sSuite
 
@@ -53,6 +54,16 @@ class MutantFinderTest extends Stryker4sSuite with LogMatchers {
       lazy val result = sut.parseFile(noFile)
 
       a[NoSuchFileException] should be thrownBy result
+    }
+
+    it("should parse a scala-3 file") {
+      import scala.meta.dialects.Scala3
+
+      val scala3DialectConfig = config.copy(scalaDialect = Scala3)
+      val sut = new MutantFinder(new MutantMatcher)(scala3DialectConfig, implicitly[Logger])
+      val file = FileUtil.getResource("scalaFiles/scala3File.scala")
+
+      noException shouldBe thrownBy(sut.parseFile(file))
     }
   }
 

--- a/core/src/test/scala/stryker4s/testutil/ExampleConfigs.scala
+++ b/core/src/test/scala/stryker4s/testutil/ExampleConfigs.scala
@@ -33,6 +33,7 @@ object ExampleConfigs {
                                      |  max-test-runner-reuse=15
                                      |  legacy-test-runner=true
                                      |  timeout=5500
+                                     |  scala-dialect="scala212"
                                      |}""".stripMargin)
 
   def empty = ConfigSource.empty
@@ -78,4 +79,9 @@ object ExampleConfigs {
                                                | timeout = 6s
                                                |}
                                                |""".stripMargin)
+
+  def scalaDialect(dialect: String) = ConfigSource.string(s"""|stryker4s {
+                                                              | scala-dialect="$dialect"
+                                                              |}
+                                                              |""".stripMargin)
 }

--- a/docs/configuration.md
+++ b/docs/configuration.md
@@ -177,6 +177,17 @@ Examples would be `sbt test`, `mvn test` or any other command to run your tests,
 
 **warning** The process runner should only be used when your specific test framework is not supported. Due to performance and predictability reasons.
 
+### `scala-dialect` ['string']
+
+**Config file:** `scala-dialect: "2.13"`  
+**Default value:** `scala3`  
+**Since:** `v0.10.1`  
+**Description:**
+
+Set the Scala dialect that should be used for parsing Scala files. The default is Scala 3 as most syntax is backwards-compatible with Scala 2. If you are running into issues with parsing older unsupported Scala syntax that we forgot about you can change this value.
+
+Valid values are Scala-versions without a patch version (`scala2.12`, `212`, `2.12`, `2`). The full list can be found [here](https://github.com/stryker-mutator/stryker4s/blob/master/core/src/main/scala/stryker4s/config/implicits/ConfigReaderImplicits.scala#L77).
+
 ## Other configuration options
 
 ### `log-level` [`string`]

